### ISSUE TITLE
Unable to connect to CloudSQL on SSL.

### ISSUE THREAD

**Issue by zaheershaik** [balderdashy/sails-mysql#227]
Hello,

I'm trying to connect to Google CloudSQL MySQL DB using sails-mysql adapter on SSL and I get the following issue. I downloaded the client certs by generating from the GC console. I'm able to connect from command line using the same certs, but not from sails. Could you please help?

Note that I've no issues connecting using the adapter without SSL.

```
mysql --ssl-ca=server-ca.pem --ssl-cert=client-cert.pem --ssl-key=client-key.pem -h HOST -u"USER" -p'PASSWORD' DB
```

```
mysqlServer: {
	module: 'sails-mysql',
	host: 'HOST',
	port: 3306,
	user: 'USER',
	password: 'PASS',
	database: 'DB',
	ssl  : {
	  key: fs.readFileSync(__dirname + '/ssl/cloudsql/client-key.pem'),
	  cert: fs.readFileSync(__dirname + '/ssl/cloudsql/client-cert.pem'),
	  ca : fs.readFileSync(__dirname + '/ssl/cloudsql/server-ca.pem')
	}
}
```

```
Sending 500 ("Server Error") response: 
 Error (E_UNKNOWN) :: Encountered an unexpected error
: Could not connect to MySQL:
Error: unable to verify the first certificate
    at afterwards (/test/node_modules/sails-mysql/lib/connections/spawn.js:72:13)
    at /test/node_modules/sails-mysql/lib/connections/spawn.js:40:7
    at Handshake.onConnect (/test/node_modules/sails-mysql/node_modules/mysql/lib/Pool.js:54:9)
    at Handshake.Sequence.end (/test/node_modules/sails-mysql/node_modules/mysql/lib/protocol/sequences/Sequence.js:96:24)
    at /test/node_modules/sails-mysql/node_modules/mysql/lib/protocol/Protocol.js:169:20
    at TLSSocket.<anonymous> (/test/node_modules/sails-mysql/node_modules/mysql/lib/Connection.js:292:7)
    at TLSSocket.emit (events.js:104:17)
    at TLSSocket._finishInit (_tls_wrap.js:460:8)
```

**Comment by rathahin** [comment]
I have the same problem problem. 

Error->[Error (E_UNKNOWN) Encountered an unexpected error] Details:  Error: Could not connect to MySQL:
Error: self signed certificate in certificate chain

I appreciate your help! Thank you very much in Advance

**Comment by cludden** [comment]
I'm also having the same issue (unable to verify the first certificate) using the three certificate files (.pem) generated via Google CloudSQL. I'm starting to think this error is coming from Google's end. I can't find the error being thrown anywhere in the adapter itself, or the node-mysql code.

**Comment by devinivy** [comment]
Please try to connect to CloudSQL using SSL manually from the same machine, and let us know what you find!  Let's figure out if this is waterline's fault before moving forward with this issue.  Thx!

**Comment by cludden** [comment]
I was able to connect successfully using the node-mysql module, which leads me to believe that the problem must lay somewhere with the adapter.

```
// test.js

var path = require('path'),
    fs = require('fs'),
    mysql = require('./node_modules/sails-mysql/node_modules/mysql');

var connection = mysql.createConnection({
    host: process.env.CLOUD_SQL_HOST,
    user: 'client',
    password: 'Tr1pl3r41nb0w',
    database: 'wccnode',
    ssl: {
        ca: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/server-ca.pem'),
        cert: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/client-cert.pem'),
        key: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/client-key.pem')
    }
});
```

**Comment by cludden** [comment]
I was able to connect successfully using the underlaying node-mysql module by itself, so now I'm beginning to think the problem must be somewhere within the sails-mysql adapter.

```
// test.js

var fs = require('fs'),
    mysql = require('./node_modules/sails-mysql/node_modules/mysql');

// this works:
var connection = mysql.createConnection({
    host: process.env.CLOUD_SQL_HOST,
    user: process.env.CLOUD_SQL_USER,
    password: process.env.CLOUD_SQL_PWD,
    database: process.env.CLOUD_SQL_DB,
    ssl: {
        ca: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/server-ca.pem'),
        cert: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/client-cert.pem'),
        key: fs.readFileSync(__dirname + '/config/_ssl/cloudsql/client-key.pem')
    }
});
```

**Comment by cludden** [comment]
I think the problem may have something to do with the fact that at some point the file buffers are getting converted into objects:

this works and comes from a console.log of the original config options:
```
{ host: ****,
  user: ****,
  password: *****,
  database: *****,
  ssl: 
   { ca: <Buffer bunch of stuff here... >,
     cert: <Buffer bunch of stuff here ... >,
     key: <Buffer bunch of stuff here... > } 
}
```

however, when I log the config options from inside node-mysql's Pool.getConnection method, the ssl certifications have been transformed into:
```
{ host: *******,
  port: 3306,
  localAddress: undefined,
  socketPath: undefined,
  user: *****,
  password: undefined,
  database: *****,
// some other stuff here
  ssl: 
   { key: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        '10': 32,
//etc
        length: 1674,
        parent: [Object] },
     cert: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        '10': 32,
//etc
        length: 1146,
        parent: [Object] },
     ca: 
      { '0': 45,
        '1': 45,
        '2': 45,
        '3': 45,
        '4': 45,
        '5': 66,
        '6': 69,
        '7': 71,
        '8': 73,
        '9': 78,
        '10': 32,
//etc
        length: 1146,
        parent: [Object] },
     rejectUnauthorized: true },
  multipleStatements: false,
  typeCast: true,
  maxPacketSize: 0,
  charsetNumber: 33,
  clientFlags: 455631 }
```

**Comment by devinivy** [comment]
Thanks!  Marking this as a bug, would love to take a peek at a PR.

**Comment by cludden** [comment]
As a follow up, the offending line is actually in waterline itself:

/lib/waterline.js
```
// Call the registerConnection method
connection._adapter.registerConnection(_.cloneDeep(config), usedSchemas, function(err) {
  if(err) return nextItem(err);
  nextItem();
});
```

_.cloneDeep(config) converts the file buffers to objects. Perhaps the waterline repo is the proper place for this issue.

**Comment by cludden** [comment]
FYI, ive opened a ticket on the waterline repo:

https://github.com/balderdashy/waterline/issues/1073

**Comment by sailsbot** [comment]
Thanks for posting, @zaheershaik.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
